Implemented validation to ensure each OTP device has a unique label

SaikrishnaPavuluri opened the pull request:
Closes #38465

<!---
Please read https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

**### Summary of Changes**
This PR enhances the OTP credential creation process by improving duplicate device validation and error handling.

Changes in **JpaUserCredentialStore** and **JpaUserFederatedStorageProvider**

1. Added duplicate credential validation for both create and update methods.

2. Ensured that users cannot register an OTP device with the same name multiple times.

3. Introduced validateDuplicateUserCredential to handle duplicate detection.

**### Enhancements in createOTPCredential (CredentialHelper.java)**

1. Changed return type from boolean to Map<String, Boolean> for more detailed status tracking:

2. **isDeviceDuplicate:**  true if a duplicate OTP device exists.

3. **isOTPCreationSuccess:** true if OTP credential creation was successful.

Handled **ModelDuplicateException**  to prevent duplicate device registrations.

**### Updates in processAction() (UpdateTotp class)**

1. Now checks **isDeviceDuplicate** before proceeding with OTP creation.

2. If a duplicate device is detected, it shows an error message:

**"Device already exists with the same name"**

3. If OTP creation fails for other reasons, it shows:

**"Invalid TOTP"**

###  Additional Update:

1. Added duplicate and empty device name validation in **setCredentialUserLabel** method **(UserResource class)**.

2. This ensures users cannot update a **userlabel** (**i.e device name**) to a blank value or a label that already exists.



**### Feedback & Review Request**

This is my first contribution to Keycloak, and I would appreciate any feedback on the implementation, approach, or any best practices I might have missed.

**Please find the validated screenshot after implementation**

![image](https://github.com/user-attachments/assets/047be3a9-6466-4907-aebf-6d87e454fbd4)


Let me know if any improvements or additional changes are needed. Thanks! 🙌

Commit:
#38465 Implemented validation to ensure each OTP device has a unique name

Signed-off-by: Saikrishna <saikrishnap@optimeyes.ai>

Commit:
Fixes #38465: Validate duplicate and empty device names in setCredentialUserLabel (UserResource)

Signed-off-by: Saikrishna <saikrishnap@optimeyes.ai>

SaikrishnaPavuluri commented:
**Additional Update:**
1. Added duplicate and empty device name validation in **setCredentialUserLabel** method **(UserResource class)**.
2. This ensures users cannot update a **userlabel (i.e device name)** to a blank value or a label that already exists.

Please find validated screenshots:

![image](https://github.com/user-attachments/assets/64e30019-347c-4529-8869-c5f524e916d6)
![image](https://github.com/user-attachments/assets/9832b37e-cfbd-4c2d-8246-bce654e45e8b)

Thanks!

SaikrishnaPavuluri commented:
Hi @ahus1,

I've made two commits to this draft PR and would really appreciate it if you could take a look whenever you get a chance. Please let me know if you have any feedback, suggestions, or improvements you'd like to see. I'd be happy to incorporate them.

Thank you for your time

ahus1 committed:
Review

Closes #38465

Signed-off-by: Alexander Schwartz <aschwart@redhat.com>

ahus1 commented on services/src/main/java/org/keycloak/services/resources/admin/UserResource.java:
```diff
@@ -886,42 +886,49 @@ public void setCredentialUserLabel(final @PathParam("credentialId") String crede
             if (auth.users().canQuery()) throw new NotFoundException("Credential not found");
             else throw new ForbiddenException();
         }
+
+        if (userLabel == null || userLabel.trim().isEmpty()) {
+            throw new ErrorResponseException("missingDeviceName", "Please specify the device name", Status.BAD_REQUEST);
+        }
+
```
As this is also used for all credentials including passwords, having an error message that is specific for a device name isn't a good fit here.

SaikrishnaPavuluri replied:
Hi @ahus1, I have updated the error to:

`throw new ErrorResponseException("missingCredentialLabel", "Credential label must not be empty", Status.BAD_REQUEST);`

Please let me know if this looks good or if you’d prefer a different message/key. Thanks!

ahus1 reviewed (changes requested):
Thank you for the pull request. I pushed one more commit to suggest  a change that doesn't make as many assumptions about the type of exceptions thrown, as there might be other exceptions appearing in other places.

I also see that you reformatted parts of the code, so the indentation has changed. Please revert those changes as they make it hard to review the changes on GitHub.

ahus1 commented:
Another hint: Please avoid to merge the main branch into the PR, instead rebase the pull request when needed. This makes it again simpler to review.

ahus1 commented:
@SaikrishnaPavuluri - please don't merge the main branch into your PR, as this makes it difficult to review. 

Rebasing should only be done when there are conflicts or a maintainer asks you to rebase. This avoid unnecessary CI runs and unnecessary notifications to maintainers.

ahus1 reviewed (changes requested):
In a previous review I commented the following, which hasn't been addressed yet: 

> I also see that you reformatted parts of the code, so the indentation has changed. Please revert those changes as they make it hard to review the changes on GitHub.

The functional changes look good, please now add a test case that validates the new behavior. Maybe you can extend an existing test, maybe you need to write a new test. Let me know if you need pointers.

Commit:
test: add tests for setCredentialUserLabel (empty, duplicate), createCredential and updateCredential (duplicate label)

Signed-off-by: Saikrishna <saikrishnap@optimeyes.ai>

Fix indentation issues

Completed review changes and reverted formatting

SaikrishnaPavuluri commented:
@ahus1 
Thanks! I've addressed the review changes.
Could you also please share some pointers on where I can add or extend the test case to validate the new behavior? That would help me get started in the right direction.

SaikrishnaPavuluri commented:
> @SaikrishnaPavuluri - please don't merge the main branch into your PR, as this makes it difficult to review.
> 
> Rebasing should only be done when there are conflicts or a maintainer asks you to rebase. This avoid unnecessary CI runs and unnecessary notifications to maintainers.

I have fixed the indentation issues and pushed the updated commit. Thanks a lot for your guidance I will be careful with formatting in future contributions to avoid this again.

SaikrishnaPavuluri commented:
> The functional changes look good, please now add a test case that validates the new behavior. Maybe you can extend an existing test, maybe you need to write a new test. Let me know if you need pointers.

Hi @ahus1 
Could you please share the pointers you mentioned yesterday for adding the test case? Appreciate your help. Thanks!

ahus1 commented:
@SaikrishnaPavuluri - I've read your comment, but allow for some more time to get back to you.

ahus1 commented:
Ok, KC 26.2 has been released, and as the dust settles, here's my reply. 

Sebastian and me recorded a session in Keycloak Hour of Code. See https://www.youtube.com/watch?v=gX9BWk-AiqE&list=PLFoHXRSurvbHNojMgLsn76-tdm13carIh&index=4&t=14s&ab_channel=AlexanderSchwartz for the video. And you find links including the script in the description of the video. 

I hope this is helpful. Please let me know if you have questions.

SaikrishnaPavuluri commented:
> Ok, KC 26.2 has been released, and as the dust settles, here's my reply.
> 
> Sebastian and me recorded a session in Keycloak Hour of Code. See https://www.youtube.com/watch?v=gX9BWk-AiqE&list=PLFoHXRSurvbHNojMgLsn76-tdm13carIh&index=4&t=14s&ab_channel=AlexanderSchwartz for the video. And you find links including the script in the description of the video.
> 
> I hope this is helpful. Please let me know if you have questions.

I will try to complete this by the end of the week. Thanks

SaikrishnaPavuluri commented:
Hi @ahus1, I have added test cases in the latest commit for **setCredentialUserLabel**, **createCredential**, and **updateCredential** to validate behavior with empty and duplicate device labels.

I also looked for a test class covering the processAction() method in UpdateTotp, but couldn’t find one. Please let me know if it exists elsewhere.

Thanks!

SaikrishnaPavuluri commented:
Hi @ahus1 
Could you please review the code for the integration tests and let me know if any changes are needed?
Thanks for your time!

ahus1 committed:
Adding a test for the AIA

Closes #38465

Signed-off-by: Alexander Schwartz <aschwart@redhat.com>

ahus1 commented:
I found a place in `AppInitiatedActionTotpSetupTest` where to add a test and pushed it to this PR. As there were several merge commits, I also rebased it for better reviewability. 

Let's look at the test results once the tests complete.

ahus1 committed:
Fixing test

Closes #38465

Signed-off-by: Alexander Schwartz <aschwart@redhat.com>

ahus1 committed:
Fixing existing tests

Closes #38465

Signed-off-by: Alexander Schwartz <aschwart@redhat.com>

ahus1 committed:
Fixing existing tests

Closes #38465

Signed-off-by: Alexander Schwartz <aschwart@redhat.com>

ahus1 committed:
Fixing existing test FlowTest.testRemoveUsedFlow

Closes #38465

Signed-off-by: Alexander Schwartz <aschwart@redhat.com>

SaikrishnaPavuluri commented:
Apologies, I made a mistake in my earlier comment. In the **UserTest class**, the method **testShouldFailToSetCredentialUserLabelWhenLabelAlreadyExists** was referring to **user-with-two-credentials**, but there's no JSON available for that user. I’ll update it to use **user-with-two-configured-otp**, which is properly defined, and push the change later.

ahus1 commented:
> Apologies, I made a mistake in my earlier comment. In the **UserTest class**, the method **testShouldFailToSetCredentialUserLabelWhenLabelAlreadyExists** was referring to **user-with-two-credentials**, but there's no JSON available for that user. I’ll update it to use **user-with-two-configured-otp**, which is properly defined, and push the change later.

This change is already part of 929daf3b61e335e69976c1ffbe372907ae3aa228

ahus1 committed:
Fixing failing test

Signed-off-by: Alexander Schwartz <aschwart@redhat.com>

ahus1 committed:
Cleanup users at the end of the test to avoid distinct user names

Signed-off-by: Alexander Schwartz <aschwart@redhat.com>

ahus1 commented:
Thank you for the code and the test, the build is now green and the tests are stable. Unfortunately the changes broke some existing tests which I didn't foresee, and it also touched some browser-only tests that are difficult to reproduce locally.

I'm now marking it ready for review and have someone else from the team look at it as well. 

When you look at my commits, please take a moment to review them. Given the documentation above, I hope it is sufficient to run a single test locally, and possibly even debug it. While running all tests locally is very time consuming and better handled by the GitHub actions, running a single test locally is a great skill that allows you to iterate faster when developing tests. If the documentation isn't sufficient, please let me know and we'll look into enhancing it.

ahus1 commented:
@mposolda - this PR looks good to me, and the error messages shown to the users when they try to have a credential with the same label are IMHO descriptive. 

As this touches the core-clients area, please give it a review as well. As this is a first-time contribution, I'm willing to give a helping hand where needed. 

Thanks!

ahus1 commented:
@SaikrishnaPavuluri - thank you for sharing this, and kudos so @srose for preparing the recordings!

mabartos commented:
@rmartinc Could you or someone from your team check this PR as @mposolda is currently out? Thanks!

rmartinc reviewed (approved):
Thanks @SaikrishnaPavuluri and @ahus1! Changes LGTM. I was thinking about adding some note in the upgrading guide. But I understand that this is more a bug than something else. Approved.
